include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1270 @@
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_differentiation_ad_ad_drivers_h
+#define dealii_differentiation_ad_ad_drivers_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/utilities.h>
+#include <deal.II/differentiation/ad/ad_number_types.h>
+#include <deal.II/differentiation/ad/ad_number_traits.h>
+#include <deal.II/differentiation/ad/adolc_number_types.h>
+#include <deal.II/differentiation/ad/sacado_number_types.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#ifdef DEAL_II_WITH_ADOLC
+
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
+#include <adolc/drivers/drivers.h>
+#include <adolc/taping.h>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
+
+#endif // DEAL_II_WITH_ADOLC
+
+#include <vector>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace Differentiation
+{
+  namespace AD
+  {
+
+    /**
+     * @addtogroup Exceptions
+     */
+    //@{
+
+    /**
+     * Exception denoting that a class requires some specialization
+     * in order to be used.
+     */
+    DeclExceptionMsg (ExcRequiresADNumberSpecialization,
+                      "This function is called in a class that is expected to be specialized "
+                      "for auto-differentiable numbers.");
+
+    /**
+     * Exception denoting that Adol-C is a required feature.
+     */
+    DeclExceptionMsg (ExcRequiresAdolC,
+                      "This function is only available if deal.II is compiled with ADOL-C.");
+
+    /**
+     * This exception is raised whenever the an auto-differentiable number does not
+     * support the required number of derivative operations
+     *
+     * The first parameter to the constructor is the number of derivative operations
+     * that it provides, and the second is the minimum number that are required.
+     * Both parameters are of type <tt>int</tt>.
+     */
+    DeclException2 (ExcSupportedDerivativeLevels,
+                    std::size_t, std::size_t,
+                    << "The number of derivative levels that this auto-differentiable number supports is "
+                    << arg1 << ", but it is required that it supports at least " << arg2 << " levels.");
+
+    //@}
+
+    /**
+     * A driver class for taped auto-differentiable numbers.
+     *
+     * It is intended that this class be specialized for the valid
+     * combinations of auto-differentiable numbers and output scalar
+     * number types.
+     *
+     * @tparam ADNumberType A type corresponding to a supported
+     *         auto-differentiable number.
+     * @tparam ScalarType A real or complex floating point number.
+     * @tparam T An arbitrary type resulting from the application of
+     *         the SFINAE idiom to selectively specialize this class.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template<typename ADNumberType, typename ScalarType, typename T = void>
+    struct TapedDrivers
```

Comment:
I may just not have read far enough down to see the specializations. But all of the member functions of this class are `static`, and none actually take any non-trivial arguments. What are they supposed to work on?

I presume that you are in essence defining an interface here, so in specializations of this class, the argument lists are going to consist of elementary data types as well (as opposed to including arguments of type `ADNumberType`). Are functions in specializations not `static` then, or how does this work? (And if there is an answer, could it be put into this class's documentation?)

Reply:
I fully understand what you find wrong with this, but for now how I can best motivate this interface is to reiterate that this the minimal ("sane") interface to a set of background operations that I have found to consistently support both ADOL-C and Sacado. The user should never end up calling any of these functions directly, and they will be used in the appropriate manner by the high-level helper classes.

For Sacado types, we set up, manipulate and extract data directly from each number. The one exception is [here](https://github.com/dealii/dealii/pull/5499/files/26e54567f92db68f410029ef80e7af2cade1d0d2#diff-42fec071a85b1a667a206111406371e6R905) - a call to a global function.

For ADOL-C, this is not the case. They have some global variables lying around that are directly manipulated by some global functions. See, for example, [here (taped mode)](https://github.com/dealii/dealii/pull/5499/files/26e54567f92db68f410029ef80e7af2cade1d0d2#diff-42fec071a85b1a667a206111406371e6R495) and and [here (tapeless mode)](https://github.com/dealii/dealii/pull/5499/files/26e54567f92db68f410029ef80e7af2cade1d0d2#diff-42fec071a85b1a667a206111406371e6R955). In my mind there are no "elementary" objects to manipulate. They're managed by ADOL-C itself. (You will see later that I have put in place many things to manage their management structures :-)

So in summary, I need some specialisations for some functions based on an AD number type. In some cases these functions don't actually operate on the numbers themselves (rather, they configure the environment) so it would be disingenuous to pass in an AD number as a parameter when its not actually being manipulated. Specialising functions through the struct + static function idiom seemed the best compromise for all of the use cases for the "driver" functions.

The hope is that being able to write code like
```c++
template<int dim, enum AD::NumberTypes ADNumberTypeCode, typename ScalarType>
    Vector<...>
    ADHelperScalarFunction<...>::compute_gradient()
{
 if (ADNumberTraits<ad_type>::is_taped == true)
        {
          Assert(this->active_tape()!=this->invalid_tape_index,
                 ExcMessage("Invalid tape index"));
          Assert(this->is_recording() == false,
                 ExcMessage("Cannot compute gradient while tape is being recorded."));
          Assert(this->independent_variable_values.size() == this->n_independent_variables(),
                 ExcDimensionMismatch(this->independent_variable_values.size(),this->n_independent_variables()));

          TapedDrivers<ad_type,scalar_type>::gradient(
            this->active_tape(),
            this->independent_variable_values,
            gradient);
        }
      else
        {
          Assert(ADNumberTraits<ad_type>::is_tapeless == true, ExcInternalError());
          Assert(this->independent_variables.size() == this->n_independent_variables(), ExcInternalError());

          TapelessDrivers<ad_type,scalar_type>::gradient(
            this->independent_variables,
            this->dependent_variables,
            gradient);
        }
}
```
in the helper classes makes obfuscations like
```c++
    template<int dim, enum AD::NumberTypes ADNumberTypeCode, typename ScalarType>
    void
    ADHelperBase<...>::ADHelperBase(
      const unsigned int n_independent_variables,
      const unsigned int n_dependent_variables)
      : ...
    {
      // Tapeless mode must be configured before any active live
      // variables are created.
      if (ADNumberTraits<ad_type>::is_tapeless == true)
        {
          TapelessDrivers<ad_type,scalar_type>::initialize(n_independent_variables);
        }
...
}
```
tolerable.

Reply:
@bangerth Do you have any suggestions or recommendations here? I'll rebase and address all of the other comments in the mean time, but it would be great to some idea as to whether I need to address this point and that below...

Reply:
I think I understand now what you are trying to do, though I'll admit that it is significantly above my head :-)

The example you gave made sense to me: In essence, you're trying to overload functions based on type, but because some of them manipulate a global environment (not an object passed in), you can't do it via function overloading -- so you do it through specialization of classes. That's a valid approach.